docs(auth): Auth API Definition (#159)
carbon-design-system/carbon-platform: docs/services-web-app.md

1. Select all repo scopes
1. Copy the `/services/web-app/.env.example` file and rename to `/services/web-app/.env.local`
1. Add your token to that new `.env.local` file

## Dependencies Set up

### Run Mode

The app uses run-mode to make decisions that are environment-dependant (cookies, next-server...).
This package configuration is also needed for proper functionality of the Auth Package.

Make sure required environment variables are configured for proper functionality of this package.
(See [run-mode package](./packages-api.md#run-mode) for details on how to set up the run-mode
package)

> _Note_: this package requires different values depending on environemnt, make sure to update these
> when running on dev/test/production mode
> 

Reviewer: ```suggestion
```

### Auth

This service depends on the auth package; Make sure required environment variables are configured
for proper functionality of this package. (See [auth package](./packages-api.md#auth) for details on
how to set up the auth package)

> _Note_: this package requires different environment vairables and values depending on environemnt,
> make sure to update these when running on dev/test/production mode
> 

Reviewer: ```suggestion
```

## Adding Local Certificates

> The application must run on https for IBMid authentication to work properly. For such purposes,
> local certificates must be generated to run on development environment. You may use
> [mkcert](https://github.com/FiloSottile/mkcert#installation) tool for this. With the tool
> downloaded, the certificates will be automatically generated next time you
> [run the app securely](#running-app-securely)

Reviewer: ```suggestion
In PROD run mode, the application must run on https for IBMid authentication to work properly. For such purposes,
local certificates must be generated to use authentication in the web-app when running locally in PROD mode.
The [mkcert](https://github.com/FiloSottile/mkcert#installation) tool can help generate these certificates and is used implicitly by the `srart-secure` node script of the web-app. With the tool downloaded, the certificates will be automatically generated the next time you
[run the app securely](#running-app-securely).
```

Reviewer: Actually question about this.. Did we ever get to the point where you can use auth in DEV mode without having to have certificates set up? Like do we have the ability to "fake" a login? If not, we should make an issue and aim to address that pretty soon

Author: tracking in #212

## Running App Securely

To run the app on https, it must be served from [server.js](../services/web-app/server.js). To do
this, run the following command from the [web-app's directory](../services/web-app):

`npm run start-secure`

## Protecting a Route

### Protecting Server-Side Rendered Pages

In order to require authentication before a user can access a server-side rendered page use the

> `getPropsWithAuth` utlity function to wrap the pages' _getServerSideProps_ function and wrap the

Reviewer: ```suggestion
`getPropsWithAuth` utlity function to wrap the pages' `getServerSideProps` function and wrap the
```

pages' content inside the `RequireAuth` component.

#### getPropsWithAuth

`getPropsWithAuth` expects an `authorizationChecker` function that receives the server side context
as a param and must return a boolean value indicating whether the user is authorized to view the
content or not:

> ```

Reviewer: ```suggestion
```js
```

import { getPropsWithAuth } from '@/utils/getPropsWithAuth'
import { retrieveUser } from '@/utils/retrieveUser'

> ...

Reviewer: ```suggestion
// ...
```

// Your custom authorization logic here, this one considers the user as authorized if it's email address ends in "ibm.com"
const isValidIbmUser = async (context) => {
  const user = await retrieveUser(context)
  if (user) {

>     return user.email?.endsWith('ibm.com')

Reviewer: I just realized we have a security hole here, lol. If I went and registered `iamnotreallyibm.com`, this check would still pass. So we should adjust this to check for either `.ibm.com` or `@ibm.com` as the suffix.

Author: tracked on #210 , fixed on f89271b97dbad6085201d96f8af4f799fedaca2d and updated docs accordingly

  }
  return false
}

> ...

Reviewer: ```suggestion
// ...
```

export const getServerSideProps = getPropsWithAuth(isValidIbmUser, async (/* context */) => {
  // Your normal `getServerSideProps` code here
  return {
    props: {
      // Your custom props here
    }
  }
})
```

Some `authorizationChecker` functions that serve common purposes are exported from

> [utils/auth-checkers](..services/web-app/utils/auth-checkers):

Reviewer: ```suggestion
`services/web-app/utils/auth-checkers`:
```
Since this is out of band of the docs, it's probably best not to link to it directly

- [https://localhost/samples/protectedPageWithSSRDynamicAuth?host=github.ibm.com](https://localhost/samples/protectedPageWithSSRDynamicAuth?host=github.ibm.com)
- [https://localhost/samples/protectedPageWithSSRDynamicAuth?host=github.ibm.com&repo=internal-stuff](https://localhost/samples/protectedPageWithSSRDynamicAuth?host=github.ibm.com&repo=internal-stuff)
- [https://localhost/samples/protectedPageWithSSRDynamicAuth?host=github.ibm.com&repo=private-stuff](https://localhost/samples/protectedPageWithSSRDynamicAuth?host=github.ibm.com&repo=private-stuff)

### Protecting Static Pages

There is no reusable strategy to guarantee authentication before accessing Statically Generated
Pages; Each page is responisble for handling authentication and authorization at a component-level
(e.g., fetching user , waiting for user to load and then displaying content). See "Authenticating
Statically Generated Pages" section in
[NextJs Authentication Docs](https://nextjs.org/docs/authentication) You may make use of the
`useAuth()` hook and `RequireAuth` component in your implementations:

> ```

Reviewer: ```suggestion
```js
```

  import { useAuth } from 'contexts/auth'
  import { useEffect } from 'react'

  import RequireAuth from '@/components/auth/require-auth'

  import FourOhFour from '../404'

  const ProtectedStaticPage = () => {
    const { isAuthenticated, loading, user } = useAuth()

    useEffect(() => {
      if (!loading && isAuthenticated) {
        // fetch protected data here
      }
    }, [loading, isAuthenticated])

    return loading ? null : (
      // show page if user is authenticated and email ends with ibm.com, else show 404 page
      <RequireAuth
        fallback={FourOhFour}
        isAuthorized={isAuthenticated && user?.email?.endsWith('ibm.com')}
      >
        <>
          // Your Page Content Here
        </>
      </RequireAuth>
    )
  }

>   // do NOT fetch protected data here

Reviewer: You can even go a step further here and mention that the reason you can't fetch protected data here is because there's no logged in user available at "build time" when these pages are generated; they're performed automatically by the server at the server's discretion.

  export async function getStaticProps(/* context */) {
    return {
      props: {} // will be passed to the page component as props
    }
  }

  export default ProtectedStaticPage
```

Note: DO NOT SERVE/FETCH PROTECTED INFO THROUGH `getStaticProps`, as these will be served

> automatically when the route is accessed regardless of authentication.

Reviewer: 👍 🔥

For a working example of a protected static page, (run the app securely)[#running-app-securely] and
visit [https://localhost/samples/protectedStaticPage](https://localhost/samples/protectedStaticPage)

## Retrieving User's Data

### Server Side Rendered Pages

> #### retrieveUser

Reviewer: ```suggestion
#### `retrieveUser`
```

The `retrieveUser` function will return the current user instance or null if user is not logged in;
note this function is asynchronous. This function can only be called server-side (i.e., in the
context of getServerSideProps):

> ```

Reviewer: ```suggestion
```js
```

import { retrieveUser } from '@/utils/retrieveUser'

> ...

Reviewer: ```suggestion
// ...
```

export const getServerSideProps = async (/* context */) => {
  const user = await retrieveUser(context)
  // Your normal `getServerSideProps` code here
  return {
    props: {
      user
      // Your custom props here
    }
  }
}
```

_Note_: if you are wrapping your `getServerSideProps` with `getPropsWithAuth` and your
`authorizationChecker` function calls `retrieveUser()`, the obtained user (if any) will be injected
into the pages's props; you can access it on `props.user`

> ### UseAuth() Hook

Reviewer: ```suggestion
### `useAuth`
```

> From a React Component, you can access the user's information via the `user` property in the

Reviewer: ```suggestion
From a React Component, you can access the user's information via the `user` property returned by
```

`useAuth` hook. This hook also exposes other info such as:

> - user: object containing user's data
> - loading: boolean indicating whether user is still being retrieved or not
> - isAuthenticated: booleand indicating whether the user has correctly authenticated or not
> - login: function to login user
> - logout: function to logout user

Reviewer: ```suggestion
- `user`: object containing user's data
- `loading`: boolean indicating whether user is still being retrieved or not
- `isAuthenticated`: booleand indicating whether the user has correctly authenticated or not
- `login`: function to login user
- `logout`: function to logout user
```

```

>  import { useAuth } from 'contexts/auth'

Reviewer: See comment on the feat(auth) PR about this. Might need a tweak depending on the result of that.

 ...
 const { isAuthenticated, isLoading, user, login, logout } = useAuth()
```

_Note:_ Keep in mind in some instances the user might still be loading (being retrieved) or might
not exist yet (user hasn't authenticated) when your component is trying to access it. Leverage the
`isLoading` and `isAuthenticated` properties for correct assertions

### API Call

If necessary, you can make a fetch request to '/api/user' which will return the user details or a
404 status code if no user has been found:

```
 const userResponse = await fetch('/api/user')
      if (userResponse.ok) {
        const user = await userResponse.json()
      }
```

expect the user response to look like this:

> ```
> {
>   "name":"Jane Doe",
>   "email": "jane.doe@emaildomain.com"
>   // Other User Properties
>   ...
>   }
> ```

Reviewer: ```suggestion
```jsonc
{
  "name":"Jane Doe",
  "email": "jane.doe@emaildomain.com"
  // ...Other User Properties
}
```
```

1. Select all repo scopes
1. Copy the `/services/web-app/.env.example` file and rename to `/services/web-app/.env.local`
1. Add your token to that new `.env.local` file

### IBMid Authentication Variables

**_Prerequisite:_** `.env.local` file for local development (see
[Setting up local environment variables](#setting-up-local-environment-variables))

Set the following variable values in `.env.local` for correct IBMid authentication flow:

- IBM_VERIFY_CLIENT_ID=[client id tied to App registration on SSO provisioner (get this from dev
  team)]
- IBM_VERIFY_CLIENT_SECRET=[client secret tied to App registration on SSO provisioner (get this from

>   dev team)]

Reviewer: I think it'd be better to keep these docs all in the auth service and then just point from here to that doc regarding the setup you need to do from an auth perspective.

Something like:
> Setting up local environment variables:
> ### Auth:
> This service depends on the auth package. See (insert some link here) for details on how to set up the auth package.
> ### Some other dep:
> This services depends on abcdefg. See (...) for details on how to set that up

Author: Done!
